### ISSUE TITLE
[distributed elastic] How to tolerate agent failures with etcd rendezvous backend?

### ISSUE THREAD

**Issue by gaocegege** [pytorch/pytorch#67616]
## ❓ Questions and Help

The training job will fail if one of the agents fails. Other agents return RendezvousClosedError. Can we tolerate agent failures? I did not find the doc about it.

I'd appreciate it if anyone could help me.

Thanks :clinking_glasses: :beers:

```
  File "/home/gaocegege/applications/miniconda3/envs/pytorch-dev/lib/python3.9/site-packages/torch/distributed/elastic/metrics/api.py", line 125, in wrapper
    result = f(*args, **kwargs)
  File "/home/gaocegege/applications/miniconda3/envs/pytorch-dev/lib/python3.9/site-packages/torch/distributed/elastic/agent/server/api.py", line 678, in _initialize_workers
    self._rendezvous(worker_group)
  File "/home/gaocegege/applications/miniconda3/envs/pytorch-dev/lib/python3.9/site-packages/torch/distributed/elastic/metrics/api.py", line 125, in wrapper
    result = f(*args, **kwargs)
  File "/home/gaocegege/applications/miniconda3/envs/pytorch-dev/lib/python3.9/site-packages/torch/distributed/elastic/agent/server/api.py", line 538, in _rendezvous
    store, group_rank, group_world_size = spec.rdzv_handler.next_rendezvous()
  File "/home/gaocegege/applications/miniconda3/envs/pytorch-dev/lib/python3.9/site-packages/torch/distributed/elastic/rendezvous/etcd_rendezvous.py", line 152, in next_rendezvous
    rdzv_version, rank, world_size = self._rdzv_impl.rendezvous_barrier()
  File "/home/gaocegege/applications/miniconda3/envs/pytorch-dev/lib/python3.9/site-packages/torch/distributed/elastic/rendezvous/etcd_rendezvous.py", line 284, in rendezvous_barrier
    return self.init_phase()
  File "/home/gaocegege/applications/miniconda3/envs/pytorch-dev/lib/python3.9/site-packages/torch/distributed/elastic/rendezvous/etcd_rendezvous.py", line 346, in init_phase
    raise RendezvousClosedError()
torch.distributed.elastic.rendezvous.api.RendezvousClosedError
```

**Comment by gaocegege** [comment]
oncall: elastic

**Comment by gaocegege** [comment]
cc @H-Huang

**Comment by kiukchung** [comment]
Hi @gaocegege, the semantics of rendezvous are documented [here](https://pytorch.org/docs/stable/elastic/rendezvous.html). Note that by design neither agent nor trainer rank is "stable" between re-rendezvous:

![image](https://user-images.githubusercontent.com/43595115/139701452-a6c5cb55-74d1-47ad-8b21-91c3759612eb.png)

Practically speaking, since we stride the trainer rank by `nproc_per_node`, this means that (for the homogeneous DDP case) agent `k` will create trainers `k ~ k+(nproc_per_node - 1)` (e.g. when `nproc_per_node=8`, agent 0 will create trainers `0,1,2,3,4,5,6,7`). 

torchelastic deals with agent failures the same way it deals with trainer failures. An agent failure is the same as `nproc_per_node` trainer failures. Now an agent failure is typically equal to a node/container failure and while torchelastic can "elastically" handle these types of failures, your scheduler may not. For example, if you are using gang scheduling on kubernetes (with Volcano) ANY pod failure is considered a job failure and the scheduler retries the entire job according to a user-defined policy and max # restarts. 

So in short, yes torchelastic deals with agent failures so long as the surviving/replaced capacity is between `min ~ max`, but to get true elasticity you'll need the scheduler to support "elastic gang" scheduling or simply use non-gang scheduling.

**Comment by gaocegege** [comment]
@kiukchung Thanks for your reply. Do we need to set some CLI arguments to support agent failure toleration? I got the error `RendezvousClosedError` from the other agents when killing one agent.

**Comment by gaocegege** [comment]
The agent will set the rdzv to `closed` when it is killed. Then other agents cannot get the next rendezvous because it is already in `closed` state. Thus other agents cannot run. I am not sure how to tolerate such failure now. Is there anything I misunderstood?

```python
    def shutdown(self) -> bool:
        try:
            self.set_closed()
            return True
        except BaseException as e:
            log.warning(f"Shutdown failed. Error occurred: {str(e)}")
            return False

    def set_closed(self):
        """
        Mark rendezvous 'closed' for current run_id, which is used to signal other
        participants to not attempt to perform (re-)rendezvous. This is useful
        when one of the workers decides the job is complete.
        """
        while True:
            active_version, state = self.get_rdzv_state()

            if state["status"] == "closed":
                # Already closed by someone else.
                return

            state["status"] = "closed"
```

**Comment by kiukchung** [comment]
How is the agent being “killed”? Can you link the code where you see the rendezvous.shutdown() method being called when the agent is killed prematurely?

**Comment by gaocegege** [comment]
```
WARNING:torch.distributed.elastic.agent.server.api:Received 2 death signal, shutting down workers
WARNING:torch.distributed.elastic.multiprocessing.api:Sending process 1375857 closing signal SIGINT
```

The agent received a signal, and the rdzv handler shutdown here https://github.com/pytorch/pytorch/blob/cd51d2a3ecc8ac579bee910f6bafe41a4c41ca80/torch/distributed/launcher/api.py#L274

**Comment by gaocegege** [comment]
I used ctrl-C and kill command (signal 15) to kill the agent and got the same error.

Then I am using signal 9, the agent works well. I think term signal is handled by the agent, then the rdzv is closed. Thanks for your reply.

**Comment by kiukchung** [comment]
hmm interesting, I have to try a few things out but it looks like on SIGINT (ctrl + c), the python finally block runs. It’s surprising since I know for a fact that SIGTERM and SIGKILL both just crash the program (without the guarantee that the finally block will run, unless you have some special termination handler for SIGTERM. For SIGKILL all bets are off and the program is just killed). I would’ve thought a SIGINT (it being a single and not an exception) would’ve not triggered the finally block to actually run. 

But yea, in short the rendezvous.shutdown() is not meant to be called on agent failures.

**Comment by kiukchung** [comment]
well what do you know: https://vorpus.org/blog/control-c-handling-in-python-and-trio/

a ctrl +c in python produces a KeyboardInterrupt exception which is handled by the try-catch-finally. Let me try to repro this tomorrow and report back. I remember recording a demo for scale downs last year (when elastic was in the pytorch/elastic repo prior to the upstream) and iirc i used ctrl + c to kill one of the agents. I have to take a look at the demo again to see what was different in that setup.

**Comment by gaocegege** [comment]
OK, Thanks for your time.

**Comment by gaocegege** [comment]
> It’s surprising since I know for a fact that SIGTERM and SIGKILL both just crash the program (without the guarantee that the finally block will run, unless you have some special termination handler for SIGTERM. For SIGKILL all bets are off and the program is just killed).


`kubectl delete` command sends a SIGTERM signal to the container. I think we should avoid running shutdown with this signal since the users or the [horizontalPodAutoscaler](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/) may delete the pod to scale down the job.

**Comment by gaocegege** [comment]
There's another problem with agent failures. Imagine that we have two agents with nprocpernode = 1. I killed agent-0, the agent-1 still works with rendezvous version 1.

Then I created a new agent-2, the agent destroyed version 1 because it is incomplete. It created a new version 2 and began the training.

At this time, agent-1 works at version 1, agent-2 works at version 2. Is it a bug or feature?

**Comment by kiukchung** [comment]
sounds like a bug to me, the expected sequence is:

1. rdzv version 0: agent-0 (alive), agent-1 (alive)
2. rdzv version 1: agent-0 (killed), agent-1 (alive)
3. rdzv version 2: agent-0 (restarted), agent-1 (alive)

Are you trying this with rdzv_backend=c10d or etcd?

**Comment by aivanou** [comment]
Assigned this issue to myself, will try to repro and if it is a bug, will submit a pr to resolve this

**Comment by gaocegege** [comment]
> Are you trying this with rdzv_backend=c10d or etcd?

I will try to give you a reproducible case here. I used etcd.

**Comment by gaocegege** [comment]
> > It’s surprising since I know for a fact that SIGTERM and SIGKILL both just crash the program (without the guarantee that the finally block will run, unless you have some special termination handler for SIGTERM. For SIGKILL all bets are off and the program is just killed).
> 
> `kubectl delete` command sends a SIGTERM signal to the container. I think we should avoid running shutdown with this signal since the users or the [horizontalPodAutoscaler](https://kubernetes.io/docs/tasks/run-application/horizontal-pod-autoscale/) may delete the pod to scale down the job.
> 
> https://github.com/pytorch/pytorch/blob/cd51d2a3ecc8ac579bee910f6bafe41a4c41ca80/torch/distributed/elastic/multiprocessing/api.py#L233

BTW, What do you think about it?

**Comment by gaocegege** [comment]
> Assigned this issue to myself, will try to repro and if it is a bug, will submit a pr to resolve this

@aivanou You can reproduce it with imagenet example here https://github.com/gaocegege/tf-operator/tree/pytorchelastic/examples/pytorch/elastic/imagenet or https://github.com/pytorch/elastic/blob/master/examples/imagenet/main.py:

```
# Terminal 1
$ python -m torch.distributed.run --rdzv_id=1 --max_restarts=100  --rdzv_backend=etcd --rdzv_endpoint=127.0.0.1:2379 --nnodes=1:2 --nproc_per_node=1 --monitor_interval 3 ./examples/pytorch/elastic/imagenet/imagenet.py ./examples/pytorch/elastic/imagenet/data/tiny-imagenet-200
# Terminal 2
$ python -m torch.distributed.run --rdzv_id=1 --max_restarts=100  --rdzv_backend=etcd --rdzv_endpoint=127.0.0.1:2379 --nnodes=1:2 --nproc_per_node=1 --monitor_interval 3 ./examples/pytorch/elastic/imagenet/imagenet.py ./examples/pytorch/elastic/imagenet/data/tiny-imagenet-200
```

Then you send the SIGKILL to the process in terminal 2, then you rerun the command to create a new agent. You can see that the two agents use different rdzv versions.

**Comment by kiukchung** [comment]
So I was able to repro this issue and can validate that this is indeed a bug (as a side note, when I recorded that demo this bug was NOT there so this is something new). Details of the bug can be found here: https://github.com/pytorch/pytorch/issues/67742. Please track the progress of the fix there.

**Comment by gaocegege** [comment]
> So I was able to repro this issue and can validate that this is indeed a bug (as a side note, when I recorded that demo this bug was NOT there so this is something new). Details of the bug can be found here: #67742. Please track the progress of the fix there.

Sure, thanks!

**Comment by gaocegege** [comment]
@aivanou Hi

Did you reproduce it successfully?

**Comment by kiukchung** [comment]
hey @gaocegege i was able to repro this, filed a bug report + have a fix PR (see the linked issue)  Was there another issue?

**Comment by gaocegege** [comment]
> There's another problem with agent failures. Imagine that we have two agents with nprocpernode = 1. I killed agent-0, the agent-1 still works with rendezvous version 1.
> 
> Then I created a new agent-2, the agent destroyed version 1 because it is incomplete. It created a new version 2 and began the training.
> 
> At this time, agent-1 works at version 1, agent-2 works at version 2. Is it a bug or feature?

@kiukchung Hi I mean this issue.

**Comment by gaocegege** [comment]
repro process:

> You can reproduce it with imagenet example here https://github.com/gaocegege/tf-operator/tree/pytorchelastic/examples/pytorch/elastic/imagenet or https://github.com/pytorch/elastic/blob/master/examples/imagenet/main.py:
> 
> ```
> # Terminal 1
> $ python -m torch.distributed.run --rdzv_id=1 --max_restarts=100  --rdzv_backend=etcd --rdzv_endpoint=127.0.0.1:2379 --nnodes=1:2 --nproc_per_node=1 --monitor_interval 3 ./examples/pytorch/elastic/imagenet/imagenet.py ./examples/pytorch/elastic/imagenet/data/tiny-imagenet-200
> # Terminal 2
> $ python -m torch.distributed.run --rdzv_id=1 --max_restarts=100  --rdzv_backend=etcd --rdzv_endpoint=127.0.0.1:2379 --nnodes=1:2 --nproc_per_node=1 --monitor_interval 3 ./examples/pytorch/elastic/imagenet/imagenet.py ./examples/pytorch/elastic/imagenet/data/tiny-imagenet-200
> ```
> 
> Then you send the SIGKILL to the process in terminal 2, then you rerun the command to create a new agent. You can see that the two agents use different rdzv versions.

**Comment by kiukchung** [comment]
when you SIGKILL one of the agents, do you see the workers on the surviving agents get restarted?

**Comment by gaocegege** [comment]
No, it did not restart. I think it is why they kept two different versions.

**Comment by kiukchung** [comment]
ddp should’ve failed if the rank is not full (eg when you lose a worker or agent). So yes, this is actually a known problem where elastic currently relies on the property of ddp that when any worker fails, failure is induced in all workers.

**Comment by gaocegege** [comment]
I am not sure if I get it. It works well with c10d backend. When the agent and trainer failed, other agents catch the connection reset error in the trainer process and re-rdzv to setup a new worker group.

It does not work when I use etcd backend. Is it a bug in etcd backend or DDP?

**Comment by kiukchung** [comment]
I would first investigate why in a multi node scenario a worker failure in node A does not cause workers on node B to also crash (they should’ve) might be an issue with a setting to DDP (maybe the gloo/nccl timeout is set too long so you are restarting the agent before giving a chance for the workers to fail). Or a bug in the example.

**Comment by gaocegege** [comment]
Gotcha, thanks! I will try to figure it out too. It makes us hard to use ETCD.

**Comment by kiukchung** [comment]
so interestingly this should work if you set nnodes=2:2 and also if you set nnodes=2:3. 

nnodes=1:2 is a really interesting edge case. Basically what is happening is that node A is killed, the workers on node B don’t crash (something to investigate) and when you restart nodeA, because min nodes is 1, node A will rendezvous successfully and go on to train with a single node. 

Had nnodes=2:3 and you killed and restarted nodeA, even if the workers on nodeB and nodeC haven’t crashed, when nodeA is restarted, nodeA can’t complete the rdzv because min nodes is 2 and will wait to be admitted. Then the monitoring loop on nodeB and nodeC will kick in which checks whether there are nodes waiting to be admitted and if so will tear down their local workers and restart.

the logic i described above is here: https://github.com/pytorch/pytorch/blob/cd51d2a3ecc8ac579bee910f6bafe41a4c41ca80/torch/distributed/elastic/agent/server/api.py#L874

**Comment by kiukchung** [comment]
so to make nnodes=1:2 work (even when the trainers don’t crash) you have to set the —rdzv_config last_call_timeout=60 (or at least 2x the monitor interval - which defaults to 30)

last_call_timeout is a config to control scaling thrashes and it forces nodeA to wait for at least 60seconds before considering the rendezvous version to be closed and going about it’s business. Since the monitor interval is 30seconds, nodeB will have a chance to see that nodeA is waiting to be admitted and will restart its workers. 

https://github.com/pytorch/pytorch/blob/cd51d2a3ecc8ac579bee910f6bafe41a4c41ca80/torch/distributed/elastic/rendezvous/etcd_rendezvous.py#L127

**Comment by gaocegege** [comment]
Gotcha, thanks for your explanation.

**Comment by kiukchung** [comment]
oh ha what do you know. this issue is actually a TODO right here: https://github.com/pytorch/pytorch/blob/cd51d2a3ecc8ac579bee910f6bafe41a4c41ca80/torch/distributed/elastic/agent/server/api.py#L674

**Comment by kiukchung** [comment]
Just to close on this loop so this works (notice the `last_call_timeout=10` and `--monitor_interval 5`):

```
# start agent 0 
$ LOGLEVEL=INFO torchrun --rdzv_backend c10d --rdzv_id 123 --rdzv_endpoint localhost:29500 --nnodes 1:2 --nproc_per_node 1 --monitor_interval 5 --rdzv_conf last_call_timeout=10 sleep.py

# start agent 1
$ LOGLEVEL=INFO torchrun --rdzv_backend c10d --rdzv_id 123 --rdzv_endpoint localhost:29500 --nnodes 1:2 --nproc_per_node 1 --monitor_interval 5 --rdzv_conf last_call_timeout=10 sleep.py

# kill agent 1
# (even though the workers on agent 0 don't error out - they are literally just time.sleep()'ing)
# restart agent 1
# will take up to 10 seconds for re-rendezvous to take place
```

**Comment by gaocegege** [comment]
Thanks!

Is there a PR to fix the problem, or should the users configure the arguments manually?

**Comment by kiukchung** [comment]
For now manually.

**Comment by gaocegege** [comment]
SGTM. Thanks for your reply.

**Comment by gaocegege** [comment]
Please let me know (at me @gaocegege ) if there is a PR about it, thanks!
